File: rust/simplify.rs
use crate::{VectorGeometry, VectorLineString, VectorPoint};

use libm::pow;

use alloc::vec;

impl VectorGeometry {
    /// Build sqdistances for a vector geometry
    pub fn build_sq_dists(&mut self, tolerance: f64, maxzoom: Option<u8>) {
        build_sq_dists(self, tolerance, maxzoom);
    }

    /// Simplify the geometry to have a tolerance which will be relative to the tile's zoom level.
    pub fn simplify(&mut self, tolerance: f64, zoom: u8, maxzoom: Option<u8>) {
        simplify(self, tolerance, zoom, maxzoom);
    }
}

/// build sqdistances for line vector data
pub fn build_sq_dists(geometry: &mut VectorGeometry, tolerance: f64, maxzoom: Option<u8>) {
    let maxzoom = maxzoom.unwrap_or(16);
    let tol = pow(tolerance / ((1 << maxzoom) as f64 * 4_096.), 2.);

    match geometry {
        VectorGeometry::LineString(geo) => {
            let coords = &mut geo.coordinates;
            build_sq_dist(coords, 0, coords.len() - 1, tol);
        }
        VectorGeometry::Polygon(geo) | VectorGeometry::MultiLineString(geo) => {
            let coords = &mut geo.coordinates;
            coords.iter_mut().for_each(|line| build_sq_dist(line, 0, line.len() - 1, tol));
        }
        VectorGeometry::MultiPolygon(geo) => {
            let coords = &mut geo.coordinates;
            coords.iter_mut().for_each(|polygon| {
                polygon.iter_mut().for_each(|line| build_sq_dist(line, 0, line.len() - 1, tol))
            });
        }
        _ => {}
    }
}

/// calculate simplification of line vector data using
/// optimized Douglas-Peucker algorithm
fn build_sq_dist(coords: &mut VectorLineString, first: usize, last: usize, sq_tolerance: f64) {
    coords[first].t = Some(1.);
    _build_sq_dist(coords, first, last, sq_tolerance);
    coords[last].t = Some(1.);
}

/// calculate simplification of line vector data using
/// optimized Douglas-Peucker algorithm
fn _build_sq_dist(coords: &mut VectorLineString, first: usize, last: usize, sq_tolerance: f64) {
    let mid = (last - first) >> 1;
    let mut max_sq_dist = sq_tolerance;
    let mut min_pos_to_mid = last - first;
    let mut index: Option<usize> = None;

    let VectorPoint { x: ax, y: ay, .. } = coords[first];
    let VectorPoint { x: bx, y: by, .. } = coords[last];

    let mut i = first;
    while i < last {
        let VectorPoint { x, y, .. } = coords[i];
        let d = get_sq_seg_dist(x, y, ax, ay, bx, by);

        if d > max_sq_dist {
            index = Some(i);
            max_sq_dist = d;
        } else if d == max_sq_dist {
            // a workaround to ensure we choose a pivot close to the middle of the list,
            // reducing recursion depth, for certain degenerate inputs
            let pos_to_mid = isize::abs(i as isize - mid as isize) as usize;
            if pos_to_mid < min_pos_to_mid {
                index = Some(i);
                min_pos_to_mid = pos_to_mid;
            }
        }

        i += 1;
    }

    if max_sq_dist > sq_tolerance {
        if let Some(index) = index {
            if index - first > 1 {
                _build_sq_dist(coords, first, index, sq_tolerance);
            }
            coords[index].t = Some(max_sq_dist);
            if last - index > 1 {
                _build_sq_dist(coords, index, last, sq_tolerance);
            }
        }
    }
}

/// square distance from a point to a segment
fn get_sq_seg_dist(ps: f64, pt: f64, s: f64, t: f64, bs: f64, bt: f64) -> f64 {
    let mut s = s;
    let mut t = t;
    let mut ds = bs - s;
    let mut dt = bt - t;

    if ds != 0. || dt != 0. {
        let m = ((ps - s) * ds + (pt - t) * dt) / (ds * ds + dt * dt);

        if m > 1. {
            s = bs;
            t = bt;
        } else if m > 0. {
            s += ds * m;
            t += dt * m;
        }
    }

    ds = ps - s;
    dt = pt - t;

    ds * ds + dt * dt
}

/// Simplify a vector geometry
pub fn simplify(geometry: &mut VectorGeometry, tolerance: f64, zoom: u8, maxzoom: Option<u8>) {
    let maxzoom = maxzoom.unwrap_or(16);
    let zoom_tol = if zoom >= maxzoom { 0. } else { tolerance / ((1 << zoom) as f64 * 4_096.) };
    match geometry {
        VectorGeometry::LineString(geo) => {
            geo.coordinates = simplify_line(&geo.coordinates, zoom_tol, false, false);
        }
        VectorGeometry::Polygon(geo) | VectorGeometry::MultiLineString(geo) => {
            geo.coordinates = geo
                .coordinates
                .iter()
                .map(|line| simplify_line(line, zoom_tol, false, false))
                .collect();
        }
        VectorGeometry::MultiPolygon(geo) => {
            geo.coordinates = geo
                .coordinates
                .iter()
                .map(|polygon| {
                    polygon.iter().map(|line| simplify_line(line, zoom_tol, true, true)).collect()
                })
                .collect();
        }
        _ => (),
    }
}

/// simplified a vector line
fn simplify_line(
    line: &VectorLineString,
    tolerance: f64,
    is_polygon: bool,
    is_outer: bool,
) -> VectorLineString {
    let sq_tolerance = tolerance * tolerance;
    let size = line.len() as f64;
    if tolerance > 0. && size < (if is_polygon { sq_tolerance } else { tolerance }) {
        return line.clone();
    }

    let mut ring: VectorLineString = vec![];
    for point in line {
        if tolerance == 0. || (point.t.unwrap_or(0.0)) > sq_tolerance {
            ring.push(point.clone());
        }
    }
    if is_polygon {
        rewind(&mut ring, is_outer);
    }

    ring
}

/// In place adjust the ring if necessary
pub fn rewind(ring: &mut VectorLineString, clockwise: bool) {
    let len = ring.len();
    let mut area: f64 = 0.;
    let mut i = 0;
    let mut j = len - 2;
    while i < len {
        area += (ring[i].x - ring[j].x) * (ring[i].y + ring[j].y);
        j = i;
        i += 2;
    }
    i = 0;
    if (area > 0.) == clockwise {
        let len_half = len / 2;
        while i < len_half {
            ring.swap(i, len - i - 1);
            i += 2;
        }
    }
}

File: src/simplify.ts
import {
  VectorGeometry,
  VectorLineString,
  VectorMultiLineString,
  VectorMultiPolygon,
  VectorPolygon,
} from '.';

/**
 * Builds squared distances for the vector geometry using the Douglas-Peucker algorithm.
 * @param geometry - input vector geometry
 * @param tolerance - simplification tolerance
 * @param maxzoom - max zoom level to simplify
 */
export function buildSqDists(geometry: VectorGeometry, tolerance: number, maxzoom = 16): void {
  const tol = Math.pow(tolerance / ((1 << maxzoom) * 4_096), 2);
  const { type, coordinates: coords } = geometry;
  if (type === 'LineString')
    buildSqDist(coords as VectorLineString, 0, (coords as VectorLineString).length - 1, tol);
  else if (type === 'MultiLineString')
    (coords as VectorMultiLineString).forEach((line) => buildSqDist(line, 0, line.length - 1, tol));
  else if (type === 'Polygon')
    (coords as VectorPolygon).forEach((line) => buildSqDist(line, 0, line.length - 1, tol));
  else if (type === 'MultiPolygon')
    (coords as VectorMultiPolygon).forEach((polygon) =>
      polygon.forEach((line) => buildSqDist(line, 0, line.length - 1, tol)),
    );
}

/**
 * calculate simplification of line vector data using
 * optimized Douglas-Peucker algorithm
 * @param coords - input coordinates
 * @param first - first point index
 * @param last - last points index
 * @param sqTolerance - simplification tolerance (higher means simpler)
 */
export function buildSqDist(
  coords: VectorLineString,
  first: number,
  last: number,
  sqTolerance: number,
): void {
  coords[first].t = 1;
  _buildSqDist(coords, first, last, sqTolerance);
  coords[last].t = 1;
}

/**
 * calculate simplification of line vector data using
 * optimized Douglas-Peucker algorithm
 * @param coords - input coordinates
 * @param first - first point index
 * @param last - last points index
 * @param sqTolerance - simplification tolerance (higher means simpler)
 */
function _buildSqDist(
  coords: VectorLineString,
  first: number,
  last: number,
  sqTolerance: number,
): void {
  let maxSqDist = sqTolerance;
  const mid = (last - first) >> 1;
  let minPosToMid = last - first;
  let index: undefined | number;

  const { x: as, y: at } = coords[first];
  const { x: bs, y: bt } = coords[last];

  for (let i = first; i < last; i++) {
    const { x, y } = coords[i];
    const d = getSqSegDist(x, y, as, at, bs, bt);

    if (d > maxSqDist) {
      index = i;
      maxSqDist = d;
    } else if (d === maxSqDist) {
      // a workaround to ensure we choose a pivot close to the middle of the list,
      // reducing recursion depth, for certain degenerate inputs
      const posToMid = Math.abs(i - mid);
      if (posToMid < minPosToMid) {
        index = i;
        minPosToMid = posToMid;
      }
    }
  }

  if (index !== undefined && maxSqDist > sqTolerance) {
    if (index - first > 1) _buildSqDist(coords, first, index, sqTolerance);
    coords[index].t = maxSqDist;
    if (last - index > 1) _buildSqDist(coords, index, last, sqTolerance);
  }
}

/**
 * square distance from a point to a segment
 * @param ps - the reference point x
 * @param pt - the reference point y
 * @param s - the first point x in the segment
 * @param t - the first point y in the segment
 * @param bs - the last point x in the segment
 * @param bt - the last point y in the segment
 * @returns - the square distance
 */
function getSqSegDist(
  ps: number,
  pt: number,
  s: number,
  t: number,
  bs: number,
  bt: number,
): number {
  let ds = bs - s;
  let dt = bt - t;

  if (ds !== 0 || dt !== 0) {
    const m = ((ps - s) * ds + (pt - t) * dt) / (ds * ds + dt * dt);

    if (m > 1) {
      s = bs;
      t = bt;
    } else if (m > 0) {
      s += ds * m;
      t += dt * m;
    }
  }

  ds = ps - s;
  dt = pt - t;

  return ds * ds + dt * dt;
}

/**
 * Simplifies the vector geometry based on zoom level and tolerance.
 * @param geometry - input vector geometry
 * @param tolerance - simplification tolerance
 * @param zoom - curent zoom
 * @param maxzoom - max zoom level
 */
export function simplify(geometry: VectorGeometry, tolerance: number, zoom: number, maxzoom = 16) {
  const zoomTol = zoom >= maxzoom ? 0 : tolerance / ((1 << zoom) * 4_096);
  const { type, coordinates: coords } = geometry;
  if (type === 'LineString')
    geometry.coordinates = simplifyLine(coords as VectorLineString, zoomTol, false, false);
  else if (type === 'MultiLineString')
    geometry.coordinates = (coords as VectorMultiLineString).map((line) =>
      simplifyLine(line, zoomTol, false, false),
    );
  else if (type === 'Polygon')
    geometry.coordinates = (coords as VectorPolygon).map((line, i) =>
      simplifyLine(line, zoomTol, true, i === 0),
    );
  else if (type === 'MultiPolygon')
    geometry.coordinates = (coords as VectorMultiPolygon).map((polygon) =>
      polygon.map((line, i) => simplifyLine(line, zoomTol, true, i === 0)),
    );
}

/**
 * @param line - input vector line
 * @param tolerance - simplification tolerance
 * @param isPolygon - whether the line is a polygon
 * @param isOuter - whether the line is an outer ring or inner ring (for polygons)
 * @returns - simplified line
 */
function simplifyLine(
  line: VectorLineString,
  tolerance: number,
  isPolygon: boolean,
  isOuter: boolean,
): VectorLineString {
  const sqTolerance = tolerance * tolerance;
  const size = line.length;
  if (tolerance > 0 && size < (isPolygon ? sqTolerance : tolerance)) return line;

  const ring: VectorLineString = [];
  for (const point of line) {
    if (tolerance === 0 || (point.t ?? 0) > sqTolerance) ring.push({ ...point });
  }
  if (isPolygon) rewind(ring, isOuter);

  return ring;
}

/**
 * In place adjust the ring if necessary
 * @param ring - the ring to rewind
 * @param clockwise - whether the ring needs to be clockwise
 */
export function rewind(ring: VectorLineString, clockwise: boolean): void {
  let area = 0;
  for (let i = 0, len = ring.length, j = len - 2; i < len; j = i, i += 2) {
    area += (ring[i].x - ring[j].x) * (ring[i].y + ring[j].y);
  }
  if (area > 0 === clockwise) {
    for (let i = 0, len = ring.length; i < len / 2; i += 2) {
      swapPoints(ring, i, len - i - 1);
    }
  }
}

/**
 * Only swap the x, y, and z coordinates
 * @param ring - the ring
 * @param i - i position in the ring
 * @param j - j position in the ring
 */
function swapPoints(ring: VectorLineString, i: number, j: number): void {
  const tmp = ring[i];
  ring[i] = ring[j];
  ring[j] = tmp;
}
